### DISCUSSION TITLE
Start of an update to the Performance Handling

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#580] [General]
I've started an update to the performance handling on the way to thinking about implementing banks and subdirectories.

Hopefully this will start to include the results of the discussions from https://github.com/probonopd/MiniDexed/issues/533, https://github.com/probonopd/MiniDexed/issues/571, https://github.com/probonopd/MiniDexed/pull/500, https://github.com/probonopd/MiniDexed/pull/575, https://github.com/probonopd/MiniDexed/discussions/569 (have I missed any?)

The first step is to make the numbers in the filenames match the user's idea of "performance number" over MIDI, so that is the first change, to make sure that works ok without breaking anything!

For progress, see: https://github.com/diyelectromusic/MiniDexed/tree/PerformanceUpdates

It would be really handy if this could get a bit of a shakedown.  I think I've tested it with missing numbers, duplicates, and tried load/save/delete and selecting via Program Change over MIDI.  I've added code to make the UI skip in the right places and while I'm there, and anticipating lots of scrolling, I've added wrap around for the selection of performances.

I /think/ it works for me, but could do with some more thorough checking.

If this is ok, then the next step will probably be:
1. Implement sub-directories in the performance area.
2. Add bank (subdir) selection via MIDI.
3. Add bank (subdir) selection via the UI.
4. Fix anything else this all throws up - there will be issues...
5. Update the behaviour as my interpretation of all the above (hard to follow!) discussions will almost certainly be different to yours!

But I'm not moving forward until I know the index/number handling is robust first - so if you're able to give it a go please do :)

By the way, should we have a "development" category in the discussion area for this kind of thing?

Kevin

**Comment by probonopd** [comment]
Hi Kevin, thanks for taking the initiative.

I have opened https://github.com/probonopd/MiniDexed/pull/581 for this. It will build automatically every time you make changes in your repository to the same branch.

**Comment by probonopd** [comment]
(MiniDexed freezing when changing performances with rotary encoder crashes in USB Gadget Mode)

 __Update:__ MiniDexed_2023-12-03-c6325cd.zip from the main branch also crashes in the same way, so this freeze is not related to the change discussed here. I have opened a separate ticket (#583) for it.

**Reply by diyelectromusic** [reply]
Ooh, that's an interesting one... will see if that happens for me too :)

Kevin

**Reply by diyelectromusic** [reply]
Well I can't reproduce exactly your situation, but I can send a MIDI file to MiniDexed using MidiEditor on Windows and all seems to work as I'm changing performances past Legend'83.  I have a Pi 3 in Gadget mode.

Does it happen if you go straight from the Default performance to Leged'83? If not, can you load any 8 performances and then have the 9th crash?  I guess I'm asking if it seems to be related to that specific performance or to the number of performance changes you've made...

If you use a different MIDI file does it still have the issue?

Would be interesting to enable MIDI debugging and see what the last thing received is when it crashes.

Kevin

**Reply by probonopd** [reply]
Thanks for testing. It is not specifically about Legend'83. It happens with the 9th performance I am trying to load. Next things I will test (though probably not today):
* Switching performances without playing MIDI at the same time
* The MiniDexed main branch to see if it happens there, too
* Not using HDMI

**Reply by probonopd** [reply]
MiniDexed_2023-12-03-c6325cd.zip from the main branch also crashes in the same way, so this freeze is not related to the change discussed here. I have opened a separate ticket (https://github.com/probonopd/MiniDexed/issues/583) for it.

**Comment by probonopd** [comment]
When you change performances via MIDI, the `[L]` stays at the previously loaded performance.

__Update:__ PCCH was disabled, my fault, sorry.

**Reply by Banana71** [reply]
My Zer0 2 doesn't behave like that. The status [L] behaves as usual. 

>When I select "Legend'83" it crashes and I hear a looping sound.

I can't provoke the crash either. 
However, I have an RPi Zero 2 with HD44780 LCD in USBGaget mode.

**Reply by probonopd** [reply]
Now it is getting interesting. (In case it is relevant, I am using a 0.96" OLED via i2c.)

**Reply by probonopd** [reply]
PCCH was disabled, my fault, sorry for the noise.

**Comment by probonopd** [comment]
So in summary after more testing, the performance handling seems to work exactly as intended and I was not able to find any additional issues in MiniDexed that didn't exist already before. Great work @diyelectromusic, thank you very much and thanks for your patience with me as a tester.

**Reply by diyelectromusic** [reply]
hehe - no problem!  Just glad someone is giving it a bit of a go :)

I wonder what your issue is though... I'll chew that over a little more.  Hmm.

Kevin

**Reply by diyelectromusic** [reply]
What would be really good is to do some loading/saving/deleting... booting with gaps in the numbering, too many performances, duplicate numbers, no performances at all, missing directories... that kind of thing.  Basically anything you can think of.

Then once I'm confident everything has been caught so far (i.e. it is all working as it did before when everything was simply sequential) then I can move to the next stage.  But if I've introduced some issues here - and tbh it wasn't easy to track all the bits down - then they'll be really hard to spot once subdirectories are added!

Kevin

**Reply by Piattica** [reply]
Kevin is it possible to have performance menu as first voice in menu?

**Reply by diyelectromusic** [reply]
It is possible, but I'm not going to change that behaviour as part of this update.  I think having a configurable "home" menu position would be useful.  Some may want it to be performances, some may want voices, some may want something else.  I already did a change on a  Pi V1 so that it defaults to the TG1 menu rather than the default top level so some of the required changes are part way there already.

Anyway, as I say, probably, but not part of this change for now.

**Reply by Piattica** [reply]
...having a configurable "home" menu position would be useful...

yes it would be great
